File: aspnet-in-practice/08-09/CS/CoolMVBlog/Utils/PermanentRedirectResult.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;

namespace CoolMvcBlog.Utils
{
    public class PermanentRedirectResult : ActionResult
    {
        private ActionResult innerResult;

        public PermanentRedirectResult(RedirectToRouteResult result)
        {
            if (result == null)
                throw new ArgumentNullException("result");

            this.innerResult = result;
        }

        public override void ExecuteResult(ControllerContext context)
        {
            this.innerResult.ExecuteResult(context);

            var response = context.HttpContext.Response;
            var url = response.RedirectLocation;

            response.RedirectPermanent(url, false);
        }
    }
}

File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Utils/PermanentRedirectResult.vb
Public Class PermanentRedirectResult
    Inherits ActionResult

    Private innerResult As ActionResult

    Public Sub New(ByVal result As RedirectToRouteResult)
        If result Is Nothing Then
            Throw New ArgumentNullException("result")
        End If

        Me.innerResult = result
    End Sub

    Public Overrides Sub ExecuteResult(ByVal context As ControllerContext)
        Me.innerResult.ExecuteResult(context)

        Dim response = context.HttpContext.Response
        Dim url = response.RedirectLocation

        response.RedirectPermanent(url, False)
    End Sub
End Class